File: Machine_Learning_A_Z/Part 2 - Regression/Section 5 - Multiple Linear Regression/my_multiple_linear_regression.py
# Multiple Linear Regression

# Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Import dataset
dataset = pd.read_csv('Machine Learning A-Z New/Part 2 - Regression/Section 5 - Multiple Linear Regression/50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# Encoding categorical data
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.compose import ColumnTransformer

ct = ColumnTransformer([('encoder', OneHotEncoder(), [3])], remainder='passthrough')
X = np.array(ct.fit_transform(X), dtype=np.float)

# Avoiding Dummy Variable Trap
X = X[:, 1:]

# Splitting dataset into Training and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Fitting Multiple Linear Regression to Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Predicting Test set results
y_pred = regressor.predict(X_test)

# Building a better model (Backward Elimination)
import statsmodels.api as sm
X = np.append(np.ones((50, 1)).astype(int), X, axis=1)
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0, 1, 3, 4, 5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0, 3, 4, 5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0, 3, 5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0, 3]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()


# Automatic Backward Elimination (p-values only)
def backward_elimination(x, sl):
    num_vars = len(x[0])
    for i in range(0, num_vars):
        regressor_OLS = sm.OLS(y, x).fit()
        max_var = max(regressor_OLS.pvalues).astype(float)

        if max_var > sl:
            for j in range(0, num_vars - i):
                if regressor_OLS.pvalues[j].astype(float) == max_var:
                    x = np.delete(x, j, 1)

    print(regressor_OLS.summary())
    return x


SL = 0.05
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
X_modeled = backward_elimination(X_opt, SL)

File: Machine_Learning_A_Z/Part 2 - Regression/Section 5 - Multiple Linear Regression/my_multiple_linear_regression.R
# Multiple Linear Regression

# Importing Dataset
dataset = read.csv('~/Desktop/DeepLearning/Machine Learning A-Z New/Part 2 - Regression/Section 5 - Multiple Linear Regression/50_Startups.csv')

# Encode categorical data
dataset$State = factor(dataset$State,
                       levels = c('New York', 'California', 'Florida'),
                       labels = c(1, 2, 3))

# Split the dataset into Training and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Fitting Multiple Linear Regression to Training set
regressor = lm(formula = Profit ~ .,
               data = training_set)

# Predicting Test set results
y_pred = predict(regressor, newdata = test_set)

# Optimizing model with Backward Elimination
regressor = lm(formula = Profit ~ + R.D.Spend + Administration + Marketing.Spend + State,
               data = dataset)
summary(regressor)

regressor = lm(formula = Profit ~ + R.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(regressor)

regressor = lm(formula = Profit ~ + R.D.Spend + Marketing.Spend,
               data = dataset)
summary(regressor)

regressor = lm(formula = Profit ~ + R.D.Spend,
               data = dataset)
summary(regressor)

# Automatic Backward Elimination
backwardElimination <- function(x, sl) {
  num_vars = length(x)
  for (i in c(1:num_vars)) {
    regressor = lm(formula = Profit ~ ., data = x)
    max_var = max(coef(summary(regressor))[c(2:num_vars), "Pr(>|t|)"])
    
    if (max_var > sl) {
      j = which(coef(summary(regressor))[c(2:num_vars), "Pr(>|t|)"] == max_var)
      x = x[, -j]
    }
    num_vars = num_vars - 1
  }
  return(summary(regressor))
}

SL = 0.05
dataset = dataset[, c(1, 2, 3, 4, 5)]
backwardElimination(training_set, SL)
